simulation/modules/plot.py

```diff
@@ -11,61 +13,85 @@
 @attr.s(kw_only=True)
 class PlotState(ModuleState):
     step_num: int = attr.ib(default=0)
+    time_steps: np.ndarray = attr.ib(factory=lambda: np.zeros(MAX_ARRAY_LENGTH, dtype=float))
+
+    # arrays to plot against time_steps
     fungal_burdens: np.ndarray = attr.ib(factory=lambda: np.zeros(MAX_ARRAY_LENGTH, dtype=int))
     macrophage_counts: np.ndarray = attr.ib(factory=lambda: np.zeros(MAX_ARRAY_LENGTH, dtype=int))
     neutrophil_counts: np.ndarray = attr.ib(factory=lambda: np.zeros(MAX_ARRAY_LENGTH, dtype=int))
-    time_steps: np.ndarray = attr.ib(factory=lambda: np.zeros(MAX_ARRAY_LENGTH, dtype=float))
 
 
 class Plot(Module):
     name = 'plot'
+    defaults = {'plot_output_folder': 'output/plots/'}
     StateClass = PlotState
 
+    def initialize(self, state: State):
+        self.advance(state, state.time)
+
+        return state
+
     def advance(self, state: State, previous_time: float):
-        step_num = state.plot.step_num
+        plot: PlotState = state.plot
+        step_num = plot.step_num
 
-        state.plot.time_steps[step_num] = state.time
-        state.plot.fungal_burdens[step_num] = len(state.fungus.cells.alive())
-        state.plot.macrophage_counts[step_num] = len(state.macrophage.cells.alive())
-        state.plot.neutrophil_counts[step_num] = len(state.neutrophil.cells.alive())
-        # define additional values to be plotted here
+        plot.time_steps[step_num] = state.time
+        plot.fungal_burdens[step_num] = len(state.fungus.cells.alive())
+        plot.macrophage_counts[step_num] = len(state.macrophage.cells.alive())
+        plot.neutrophil_counts[step_num] = len(state.neutrophil.cells.alive())
 
-        state.plot.step_num += 1
+        plot.step_num += 1
 
         return state
 
-    def finalize(self, state: State):
-        num_steps = state.plot.step_num
-        time_steps = state.plot.time_steps[:num_steps]
-        fungal_burdens = state.plot.fungal_burdens[:num_steps]
-        macrophage_counts = state.plot.macrophage_counts[:num_steps]
-        neutrophil_counts = state.plot.neutrophil_counts[:num_steps]
-        # define additional y_val here
+    def setup_plot(self, ylabel: str, title: str, filename: str):
+        plt.xlabel('Time Unit (Typically Hours)')
+        plt.ylabel(ylabel)
+        plt.title(title)
+        plt.grid(True)
+        plt.xlim(xmin=0)
+        plt.ylim(ymin=0)
+        plt.savefig(f'{filename}.png')
+        plt.clf()
 
+    def finalize(self, state: State):
+        plot_output_folder = self.config.get('plot_output_folder')
+        if not os.path.exists(plot_output_folder):
+            os.mkdir(plot_output_folder)
+        plot: PlotState = state.plot
+        num_steps = plot.step_num
+        time_steps = plot.time_steps[:num_steps]
         plt.figure()
```

Comment:
Reusing this shared global `Figure` instance and clearing + modifying it after each plot may be getting too complex. I think it's outside the scope of this PR, but we may be near the point that we should graduate beyond the "simple" `matplotlib.pyplot` API and into the more powerful object-oriented matplotlib API. I think this will ultimately make the code easier to work with. Here's a good starting point: https://matplotlib.org/3.2.1/api/index.html#usage-patterns